repairDuplicateVersions.js

```diff
@@ -0,0 +1,447 @@
+        if (err) {
+            return cb(err);
+        }
+        const copiedKeys = {};
+        return async.eachSeries(results.objectUrl.locationKeys, (sproxydKey, done) => {
+            if (!results.objectUrl2.locationKeys.has(sproxydKey)) {
+                // sproxyd key is not duplicated
+                return done();
+            }
+            // sproxyd key is duplicated, need to copy the data to a
+            // new key and update metadata for objectUrl
+            return copySproxydKey(objInfo.objectUrl, sproxydKey, (err, newKey) => {
+                if (err) {
+                    return done(err);
+                }
+                log.info('sproxyd key copied', {
+                    objectUrl: objInfo.objectUrl,
+                    sproxydKey,
+                    newKey,
+                });
+                copiedKeys[sproxydKey] = newKey;
+                return done();
+            });
+        }, err => {
+            if (err) {
+                return cb(err);
+            }
+            if (Object.keys(copiedKeys).length === 0) {
+                log.info('skip object already repaired', {
+                    objectUrl: objInfo.objectUrl,
+                });
+                status.objectsSkipped += 1;
+                return cb();
+            }
+            const objMD = results.objectUrl.md;
+            objMD.location.forEach(loc => {
+                if (copiedKeys[loc.key]) {
+                    // eslint-disable-next-line no-param-reassign
+                    loc.key = copiedKeys[loc.key];
+                }
+            });
+            return putObjectMetadata(objInfo.objectUrl, objMD, err => {
+                if (err) {
+                    log.error('error putting object metadata', {
+                        objectUrl: objInfo.objectUrl,
+                        error: err.message,
+                    });
+                    return cb(err);
+                }
+                log.info('repaired object metadata', {
+                    objectUrl: objInfo.objectUrl,
+                });
+                status.objectsRepaired += 1;
+                return cb();
+            });
+        });
+    });
+}
+
+function repairObjects(cb) {
+    logProgress('start repairing objects');
+    async.eachSeries(objectsToRepair, (objInfo, done) => {
+        repairObject(objInfo, err => {
+            if (err) {
+                log.error('an error occurred repairing object', {
+                    objectUrl: objInfo.objectUrl,
+                    error: err.message,
+                });
+                status.objectsErrors += 1;
+            }
+            done();
+        });
+    }, cb);
+}
+
+function main() {
+    async.series([
+        done => getSproxydAlias(done),
+        done => readVerifyLog(done),
+        done => repairObjects(done),
+    ], err => {
+        if (err) {
+            log.error('an error occurred during repair process', {
+                error: err.message,
```

Comment:
Minor: the script's output will probably not be indexed, but in case it is, it's safer to do `error: { message: err.message }` otherwise the indexes choke on mismatched types for `error` being sometimes an object and sometimes a string.